### PR TITLE
Use the selection range when resolving call hierarchy items

### LINKED ISSUE #5103: Call Hierarchy is broken
Given the following file:
```rust
fn a() {
    b()
}

fn b() {}

fn main() {
    a()
}
```

`rust-analyzer` reports no incoming calls for `b` and two outgoing calls (one to `a` and one to `b`). For `a`, it also reports one outgoing call to `b` and no incoming calls.

### PR DESCRIPTION
**Pull request by kjeremy**
Add a test in call_hierarchy that already passed and a corresponding
heavy test to test the LSP requests which exposed the issue.

Fixes #5103

### CODE REVIEW

**Comment by kjeremy** [comment]
The fun thing about call hierarchy is that it exercises the goto definition and find references machinery.

File: crates/rust-analyzer/tests/heavy_tests/main.rs [hunk 1]
```diff
@@ -676,3 +680,159 @@ pub fn foo(_input: TokenStream) -> TokenStream {
     let value = res.get("contents").unwrap().get("value").unwrap().to_string();
     assert_eq!(value, r#""```rust\nfoo::Bar\n```\n\n```rust\nfn bar()\n```""#)
 }
+
+#[test]
+fn test_issue_5103() {
```
Comment (matklad) [inline crates/rust-analyzer/tests/heavy_tests/main.rs:685]: We don't add regression slow tests, as they are slow. So, this either has to be untested, or we need to move the checks to the type level.

File: crates/rust-analyzer/src/handlers.rs [hunk 2]
```diff
@@ -1062,7 +1062,7 @@ pub(crate) fn handle_call_hierarchy_outgoing(
     let item = params.item;
 
     let doc = TextDocumentIdentifier::new(item.uri);
-    let frange = from_proto::file_range(&snap, doc, item.range)?;
+    let frange = from_proto::file_range(&snap, doc, item.selection_range)?;
```
Comment (matklad) [inline crates/rust-analyzer/src/handlers.rs:1065]: I wonder if that means that our API and/or protocol are off. Positions only make sense for the first request which initializes call hierarchy. Navingating to incoming/outgoing call should not use positions. At the same time, round-tripping some kind of persistent cookies seems awkward. 

I guess, for the time being this is a fine fix, but this needs more design long-term.

Comment (kjeremy) [reply crates/rust-analyzer/src/handlers.rs:1083]: I was thinking of reconstructing the call item and passing that down and may try something like that in a followup PR.

**Pull request merged**